feat: report download progress

vkt1414 opened the pull request:
This PR aims to address #24 and #51

there is no way to track download sequentially as s5cmd run or cp command will be locked on a thread. so, download progress is outsourced to a thread and will run simultaneously along side download thread on both series and manifest downloader

as the index only contains aws urls, when a manifest contains gcs urls, crdc series instance uuid is extracted from aws urls and queried against the aws urls in the index to get download size. For manifest download, download size is calculated first and the progress is tracked against as a whole.

manifest validator in download from manifest is offloaded to a dedicated function that will check not only the first line but every line, and if the manifest has urls from both gcp and aws to raise an exception.

s5cmd cp is replaced with sync to gracefully avoid downloading the same data again.

get functions will now return a message that data not found for the values given for a key.

queries folder is now removed as they will persist in idc-index-data

vkt1414 commented:
@jcfr whenever possible, could you help troubleshoot with the pylint error in CI? I am unable to figure what's causing the issue. Thanks!
https://github.com/ImagingDataCommons/idc-index/actions/runs/8411497038/job/23031159059?pr=56

fedorov commented:
Maybe this article can help troubleshoot pylint error 12: https://stackoverflow.com/questions/76181900/pylint-return-non-zero-code-error-only-if-errors-are-there

fedorov commented:
@jcfr sorry for the silly question - but if we are running ruff already, why do we also need to run pylint?

jcfr commented:
ruff and pylint are complementary

jcfr commented:
When first updating the CI, I added a lot "exception" related to potential issues:

https://github.com/ImagingDataCommons/idc-index/blob/1868f8a2b2ec4009c9928bca5c387ae4a57bc12f/pyproject.toml#L150-L174

https://github.com/ImagingDataCommons/idc-index/blob/1868f8a2b2ec4009c9928bca5c387ae4a57bc12f/pyproject.toml#L196-L206

To move forward, it would be sensible to:
* gradually review and address these exceptions if it applies
* address warnings associated with new code & feature

Note that in some cases, it is legitimate to have exception and not follow the recommendation associated with the warning. This should be address on a case-by-case basis

vkt1414 commented:
@fedorov I added the initial disk size as a baseline, and the download progress is now going as expected. Calculating file size may be an overhead if too many files are in a directory. I do not know if there is any better way.

I tried replacing threading with multiprocessing which seems to use Popen, it worked only on Ubuntu, but it did not work well on Windows and macOS systems, so I reverted to threading.

I can also confirm that s5cmd does not delete any existing files, so replacing cp with sync has been seamless.

Here's a run that did not work on Windows and Mac.
https://github.com/ImagingDataCommons/idc-index/actions/runs/8426861486

fedorov commented:
@vkt1414 this is the use of `Popen` I was referring to at the meeting. What is wrong with this simple approach?

<img width="1336" alt="image" src="https://github.com/ImagingDataCommons/idc-index/assets/313942/f502e3c6-3dcc-4e4c-b5be-740e2e77dda5">

vkt1414 commented:
> @vkt1414 this is the use of `Popen` I was referring to at the meeting. What is wrong with this simple approach?
> 
> <img width="1336" alt="image" src="https://github.com/ImagingDataCommons/idc-index/assets/313942/f502e3c6-3dcc-4e4c-b5be-740e2e77dda5">
> 

Thank you! Ubuntu(Linux) has no problem with any approach so far. 

Either way, I'll give this a try.

fedorov commented:
The idea is to use the simplest approach possible to improve readability and reduce maintenance.

vkt1414 commented:
> The idea is to use the simplest approach possible to improve readability and reduce maintenance.

Sure. I used Popen only this time, and this PR should now be ready for review.

fedorov commented:
@vkt1414 this PR is still marked as draft - is this right?

fedorov commented on idc_index/index.py:
```diff
@@ -341,7 +422,7 @@ def get_series_file_URLs(self, seriesInstanceUID):
 
     def get_viewer_URL(
-        self, seriesInstanceUID, studyInstanceUID=None, viewer_selector=None
+        self, seriesInstanceUID=None, studyInstanceUID=None, viewer_selector=None
```
This is potentially confusing, since it implies both series and study UIDs can be None.

vkt1414 replied:
I do not know there if there is any other way. 

I notice that it is implemented the same way in `download_from_selection` function as well, where we do not explicitly require any UIDs.
https://github.com/ImagingDataCommons/idc-index/blob/main/idc_index/index.py#L505
![image](https://github.com/ImagingDataCommons/idc-index/assets/115020590/6eb33ef3-69d9-4674-beae-3541b112fcaa)

fedorov replied:
Yes, good point. I agree there is no good solution.

fedorov commented on idc_index/index.py:
```diff
@@ -444,6 +528,254 @@ def get_viewer_URL(
+        Validates the manifest file by checking the URLs and their availability.
+        The function reads the manifest file line by line. For each line, it checks if
+        the URL is valid and accessible.
+        Uses the s5cmd to check the availability of the URLs in both AWS and GCP.
```
Is this a good idea? For large manifests, this potentially can be a lengthy operation. Did you test how long it takes? We should think how to optimize this. If both the bucket and the UUID are valid (based on the index), we can rather safely assume the URL is correct, and could avoid the costly operation.

vkt1414 replied:
After rewriting the logic, we will be using download_from_manifest only when explicitly called. Given that, I think it may be important to check if the manifest was properly generated. 
I validated the 65k series manifest from portal, and it took >30 min(and counting, will update when finished) to validate.

> If both the bucket and the UUID are valid (based on the index), we can rather safely assume the URL is correct, and could avoid the costly operation.

We do not have access to the bucket id directly. But we could simplify validation for aws URLs by simply matching the entire s3 URLs in the index. If not found, then we could do `s5cmd ls`, to see if there are in GCP. What do you think? or should we simply extract and check if crdc instance uuid is in aws_url (this step is already included in the current PR)?

vkt1414 replied:
I have a ridiculous idea that may or not be understood. But putting it out there. All of the manifest validation or sync operations are currently checked one series. I think there is a way we could check as a list of series. Here are my thoughts while commuting on mbta:



Remove commented lines from the manifest file.

Query against as if it is a text file with one column, extract crdc instance uuid from both index and left join with index using crdc instance uuid, convert to a data frame and then check if there are any empty rows in the index's crdc column.


Similarly for the sync command as well.

fedorov replied:
Yep, join is exactly what I had in mind when I referred to checking bucket and UUID based on index! :-)

vkt1414 replied:
Great.  For GCP urls, I'm checking only if crdc uuid is present in the index. We can't verify the bucket name. I think it may be a reasonable assumption. The validation at the df level for 65k series cohort is <10s now, but the sync dry run took ~12 min, so a total of ~13 min before download starts

vkt1414 commented:
> @vkt1414 this PR is still marked as draft - is this right?

I have now marked it as ready for review now.

vkt1414 commented:
> > @vkt1414 this PR is still marked as draft - is this right?
> 
> I have now marked it as ready for review now.

Changing back to draft as I think I can optimize the manifest validator and sync size calculations in seconds regardless of the manifest size. I'd be analyzing all series in the manifest, instead of one series at a time as it is currently implemented now. I'd be leveraging reading the manifest as a dataframe, followed by data wrangling operations to accomplish this.

vkt1414 commented:
here's my attempt at creating the flowchart: (it's an svg file, so we can zoom)
![flowchart](https://github.com/ImagingDataCommons/idc-index/assets/115020590/60c07c08-3d11-4de7-8bfc-570f11c2c76b)

fedorov commented:
@vkt1414 I am going over the review. Your summary of what the function does is very helpful for the review! But I really do not think it should be in the docstrings - those should contain details needed to understand what the function does. Not how it does it. I would recommend moving details of how the function is implemented into the comments within the body function and outside of the docstring.

Please wait with responding to this and with any other edits until I am done with my review!

fedorov commented on idc_index/index.py:
```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+            None
+        """
+
+        dry_run_cmd = [
```
And yes, I've just tested with a larger-size manifest, and the delay introduced by this s5cmd dry-run is absolutely unacceptable!

I understand we will not be able to know more precisely the size that needs to be downloaded, but IMHO it is most definitely not worth it to do dry run just to get the accurate size and suffer this enormous delay.

vkt1414 replied:
Please see the svg above for why s5cmd dry run is being used.

without s5cmd sync dryrun, progress bar can sometimes be misleading. For example, when the same cohort is downloaded twice to the same folder, progress bar remains at 0%. I was just trying to implement as perfectly as possible with the tools at hand, while letting the user know when the progress bar can not be accurate.

 > dry run just to get the accurate size and suffer this enormous delay.

rsync also has a cold start right when working large data? So maybe it is normal. 

Your call on progress bar being inaccurate is okay when same cohort is downloaded more than once to the same folder vs the PR implementation.

fedorov replied:
I do not think this enormous delay is justified to make the progress accurate. 

Also, I now looked at the documentation for `s5cmd --dry-run` https://github.com/peak/s5cmd?tab=readme-ov-file#dry-run, and it says that it "will output what operations will be performed without actually carrying out those operations". Now I am confused how it can help at all. It will just repeat the commands we already know from the manifest.

vkt1414 replied:
> It will just repeat the commands we already know from the manifest.
Not true when someone interrupts the download. Once interrupted, we do not know what is already downloaded and what needs to be downloaded. Same with the case, when the download is tried twice for the same cohort to the same folder.

Example 1: If a manifest is 2 GB, and the data is already downloaded. If we run the same command again, the base line now is 2 GB, and the progress bar tries to expect 4 GB in the folder. But because the data is already there, progress bar will be at 0%

Example 2: If a manifest is 2 GB, and **SOME* of the data in the manifest is already downloaded in the destination folder. For example 1 GB. If we run the same command again, the baseline for the progress bar now is  1 GB, and the progress bar tries to expect a total of 3 GB in the folder. But because the data is already there, the progressbar will be stopped at 50%.
In these two examples, we can not provide any message to the user, except for one, `destination folder is not empty, progress bar may not always be accurate.`

With s5cmd sync dryrun we will know what needs to be actually downloaded. i.e if some data is already present, s5cmd sync dryrun output will not contain the instances that are already in the destination. We can parse the stdout from dryrun command, and calculate the sync size and download just those.

I'll create another branch without using sync dryrun I'll then create a colab notebook with all the cases where progress bar can be misleading.

fedorov replied:
> With s5cmd sync dryrun we will know what needs to be actually downloaded.

If dry run just prints out the commands it would execute, it will not help understanding what has already been downloaded. There file may be zero size, or incomplete, or may have content different from remote. Communication with the server will be needed to confirm whether something was completed or not, and for that you need more than dry run. I have not run any experiments, but that's my understanding based on the documentation. So I do not think it solves the problem.

> I'll create another branch without using sync dryrun I'll then create a colab notebook with all the cases where progress bar can be misleading.

I really do not see why this should be a priority at all.

You have to weigh perfection vs impact.

If the download is partially complete, your estimate will be larger than the real size to be downloaded. So the result will be that user will in reality wait less time than projected. What is the negative impact with that from the user perspective? I see no problem. In your first example, it will jump from 0 to done very quickly, in the second example user will wait half the time. Zero negative impact IMHO.

At the same time, if we seek perfection, every user will suffer from the increased delay for every download. This is clearly a significant problem for the user. So it does not matter how dry run works. Bottom line - it increases download time by a lot for even a moderate size manifest of 65K, and it is not worth the benefit that brings (even if dry run can give us that estimate).

vkt1414 replied:
> If dry run just prints out the commands it would execute, it will not help understanding what has already been downloaded. There file may be zero size, or incomplete, or may have content different from remote. Communication with the server will be needed to confirm whether something was completed or not, and for that you need more than dry run. I have not run any experiments, but that's my understanding based on the documentation. So I do not think it solves the problem.

Unfortunately, this is not true. You are stressing only on the dry-run and not the sync. Please see the documentation for how s5cmd handles sync. Sync checks what is already in the download folder by two parameters size and time of the file modified (i.e. when IDC ingested them to AWS).
https://github.com/peak/s5cmd?tab=readme-ov-file#sync

 > In your first example, it will jump from 0 to done very quickly
 Progress bar remains at 0% not jump to 100%.
 
> At the same time, if we seek perfection, every user will suffer from the increased delay for every download. This is clearly a significant problem for the user. So it does not matter how dry run works. Bottom line - it increases download time by a lot for even a moderate size manifest of 65K, and it is not worth the benefit that brings (even if dry run can give us that estimate).

In my opinion, we should implement the best way we possibly can. For me, misleading progress bar is worse than the cold start. But if you are fine with that, I'll be too.

I can think of one more solution:
1. Offer progress bar as an optional feature. Because for cases like SlicerIDCBrowser, progress bar may not be helpful at all.
2.  If the destination folder is empty, I'll agree with you that there is no need to sync dry-run.
3. Only when the user chooses progressbar=True, and the destination folder is not empty, we can do s5cmd sync dryrun.

fedorov replied:
> For me, misleading progress bar is worse than the cold start. 

Why is that?

I had another thought about how we can do it. Why do we need to check with the server at all. We can loop just over the local destination files in the manifest and calculate their size locally. It will be most definitely faster than the current dry run! I am not sure why we didn't consider this earlier.

vkt1414 replied:
Because I know of at least three cases where progress bar will be misleading. Who knows how many other possibilities there are. I discussed two cases discussed above. Third is when there are some and not all instances are removed for a series. Because we have only series level info, we can't calculate the partial size to be downloaded. This is an inherent problem with our data model and nothing much we can do about it.


I don't understand what you mean about looping over the dicom files in the destination folder. 

As you know, manifest does not have instances info, only series folder. How can we know the size (when the folder contains for example unrelated dicom files to the manifest in the destination), what is downloaded and what is not?   

Please help me understand.

fedorov replied:
> Because I know of at least three cases where progress bar will be misleading.

I do understand that. But I meant to ask why do you consider accurate progress bar to be more important than cold start? 

> As you know, manifest does not have instances info, only series folder. 

The decision to flatten the content of the series dirs is not rooted in user needs. There is no harm (quite the opposite) to keep series-level folders at the destination. If we have the list of series folders to track, everything becomes much easier.

But let's wrap up this PR and address this separately.

vkt1414 replied:
> I do understand that. But I meant to ask why do you consider accurate progress bar to be more important than cold start?

because unless we are benchmarking or using for using it for viewer applications such as SlicerIDCBrowser, speed may not be a priority. Accuracy may be. If I'm downloading 10 TB and the download is interrupted for whatever reason and if I'm counting on the progress bar to track progress, I would appreciate the accuracy. Again, we don't have to force the progress bar at all times. We could implement it as a feature. For those who value speed over an accurate progress bar, they absolutely can disable the feature as I mentioned above.

> The decision to flatten the content of the series dirs is not rooted in user needs. There is no harm (quite the opposite) to keep series-level folders at the destination. If we have the list of series folders to track, everything becomes much easier.

This is unfortunately not foolproof. Once again if some data is already present in the destination folder even with series folder implementation, yes, we may know the size of the destination series folder is different. But we will need to think about how to address the inconsistencies. We will essentially be implementing sync logic when it is already available in s5cmd.

> But let's wrap up this PR and address this separately.

Could I know if you have any objections to this?

>1. Offer progress bar as an optional feature. Because for cases like SlicerIDCBrowser, progress bar may not be helpful at all.
> 2. If the destination folder is empty, I'll agree with you that there is no need to sync dry-run.
> 3. Only when the user chooses progressbar=True, and the destination folder is not empty, we can do s5cmd sync dryrun.

fedorov replied:
> because unless we are benchmarking or using for viewer applications such as SlicerIDCBrowser, speed may not be a priority.

In my view, speed is always the priority. One of the modes of operation for SlicerIDCBrowser is interactive downloader, so it will be relevant there as well. Based on the arguments I have seen, I would like to go ahead with prioritizing speed over accuracy of progress. If user concerns emerge over time, I am ready to revisit. If you like, we can bring this up at the weekly meeting tomorrow.

> Offer progress bar as an optional feature. Because for cases like SlicerIDCBrowser, progress bar may not be helpful at all.

This is actually a good point we did not consider earlier. Progress reporting will most definitely be very important for SlicerIDCBrowser. We need a function that will be able to help us update the progress bar in Slicer. Can you look into how progress bar is implemented in Slicer typically, and see what we might need to do to support it with this function?

> Only when the user chooses progressbar=True, and the destination folder is not empty, we can do s5cmd sync dryrun.

I think both progress bar and dryrun should be optional. Whenever we get to providing a command line download tool, progress should be enabled by default, and dryrun should be disabled by default.

vkt1414 replied:
>In my view, speed is always the priority. One of the modes of operation for SlicerIDCBrowser is interactive downloader, so it will be relevant there as well. 
This is actually a good point we did not consider earlier. Progress reporting will most definitely be very important for SlicerIDCBrowser. We need a function that will be able to help us update the progress bar in Slicer. Can you look into how progress bar is implemented in Slicer typically, and see what we might need to do to support it with this function?


I'm totally aware of need to bring progress bar to SlicerIDCBrowser. The reason why I kept saying the progress bar implemented here will be not useful in SlicerIDCBrowser is because, progress bar is displayed with qt there and idc-index's progress bar will be of no use. We will need to implement it again there.

https://github.com/PerkLab/SlicerSandbox/blob/master/LoadRemoteFile/LoadRemoteFile.py#L50

> I think both progress bar and dryrun should be optional. Whenever we get to providing a command line download tool, progress should be enabled by default, and dryrun should be disabled by default.

I do not understand what you finally want here. Please correct me if my understanding is wrong. For now, everyone is opted into progress bar and no option to select whether to enable or disable. Second do not use sync dryrun at all? Is that correct?

fedorov replied:
Ok. Sorry for the confusion, let's not increase the complexity. Forget the other points. Let's do the following:

1. Make both progress reporting and dry run configurable via function parameters.
2. Progress reporting is enabled by default
3. Dry run is disabled by default

vkt1414 replied:
All requested constraints are now met.

I am not a fan and find it unethical to provide an option to skip validating manifests. As s5cmd does not interrupt download even with invalid URLs. There is no meaningful savings of any kind by skipping. But I'll leave it as it is.

![image](https://github.com/ImagingDataCommons/idc-index/assets/115020590/64e8d53e-9cd6-4ab8-8342-377d88b17ebb)

Resolving this.

fedorov commented on idc_index/index.py:
```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+            lambda row: aws_endpoint_url
+            if row["crdc_instance_uuid_found"]
+            and row["s3_url"] == row["series_aws_url"]
+            else gcp_endpoint_url,
```
I do not want to make this assumption.

I suggest for now we simplify the logic of the endpoint identification, and keep it the same as before: check the first URL, whatever it is - AWS or GCP - assume it is the same for all URLs and proceed with the download using that endpoint.

If the URL is invalid, I hope it is possible to communicate to the user which specific URL failed when copy is aborted.

We can have an option in the function to perform validation of the manifest, to do more thorough checks, but I think in most cases those are not going to be useful, since manifest will either come from us, or from the portal.

What do you think?

vkt1414 replied:
In the main branch implementation, the manifest is generalized by just looking at the first line. There could be invalid urls from second line and no

https://github.com/ImagingDataCommons/idc-index/blob/main/idc_index/index.py#L621

Compared to that, I'd argue that ths PR implementation is more rubust, not very time expensive (~3 seconds to validate 65k series manifest). But I do realize that it is not perfect and drew the line after checking a valid crdc uuid.

It is a battle of two assumptions. Checking the first line and no error handling vs checking every line for a valid crdc uuid and raising an exception if not. If you think, we can assume that the manifest is valid by looking at just the first line, I'm fine with that and revert the changes.

fedorov replied:
My concern is that in the code highlighted, if we do not find a match, we cannot really conclude that the bucket is in GCP, and that's what the code is doing right now. I do not object to matching UUIDs (although we should check how long it takes for a 400K manifest).

It is not quite a battle of two assumption. In either case, if UUID/bucket are invalid, the download operation will abort. The difference is that currently it will abort with a message "The manifest contains URLs from both AWS and GCP. Please use only one provider", which may be rather misleading. If the UUID/bucket path is invalid, it will fail to find the file and I assume will print the corresponding message.

vkt1414 replied:
> "The manifest contains URLs from both AWS and GCP. Please use only one provider", which may be rather misleading. If the UUID/bucket path is invalid, it will fail to find the file and I assume will print the corresponding message.

This is not true completely. Yes, it is possible for this message to be incorrect but only when crdc uuid is **FOUND** but the bucket ID is invalid.

Please see this code which is run before the cited code.

```

# Check if crdc_instance_uuid is found in the index
if not all(merged_df["crdc_instance_uuid_found"]):
    missing_manifest_cp_cmds = merged_df.loc[
        ~merged_df["crdc_instance_uuid_found"], "manifest_cp_cmd"
    ]
    missing_manifest_cp_cmds_str = f"The following manifest copy commands do not have any associated series in the index: {missing_manifest_cp_cmds.tolist()}"
    raise ValueError(missing_manifest_cp_cmds_str)
```


 I';l try another visual way to explain this.

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/vamsi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/vamsi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->

</head>

<body link="#467886" vlink="#96607D">


  |   |   | End point to use
-- | -- | -- | --
  | crdc uuid match | complete s3 url match |  
Case 1 | TRUE | TRUE | AWS
Case 2 | TRUE | FALSE | GCS
Case 3 | FALSE | N/A | Raise an exception
Case 4 | TRUE | Mix of TRUE and FALSE | Raise an exception that manifest has both GCS and AWS endpoints



</body>

</html>


> In either case, if UUID/bucket are invalid, the download operation will abort.

In the main branch's implementation, download begins even with an invalid manifest, i.e until it encounters an invalid line. This PR does not begin download until the manifest is valid for aws manifests. For GCP manifests, yes, it will also download until it encounters an invalid bucket path. 

Could I know what you would like me to do, after seeing the pros and cons of the PR implementation? 

Revert to checking only the first url only or keep the PR implementation?

fedorov replied:
I think I understand what you are doing. My initial comment here was about the assumption that an unknown URL is a GCP URL, and the error message that stems from that assumption.

I suggest we keep the UUID check, and use the first URL check whenever we do not have a match with AWS URL to determine if the endpoint is GCP.

fedorov replied:
I thought about this a bit, and I think we can improve this a lot with a minimal effort.

We can add this query and the result to `idc-index-data`, and it will give us mapping between AWS and GCP buckets:

```sql
SELECT
  DISTINCT(aws_bucket),
  gcs_bucket
FROM
  `bigquery-public-data.idc_current.dicom_all`
```

Then we can maintain this table separately and use it for validating GCS URLs and also for providing GCS URLs if requested by user.

It is fine to do this later. Let's just wrap this PR first.

vkt1414 replied:
> I suggest we keep the UUID check, and use the first URL check whenever we do not have a match with AWS URL to determine if the endpoint is GCP.

Sounds good to me.


>  thought about this a bit, and I think we can improve this a lot with a minimal effort.

> We can add this query and the result to idc-index-data, and it will give us mapping between AWS and GCP buckets:

I know of a solution but you may not like it. In my opinion, there is no reason to have duplicate content in any columns. i.e we must create nesting. As I have mentioned before, users will never have to unnest. It is us, who have the expertise to unnest could take advantage, to sacrifice a little bit of time for sustainability. But that's a rant for another day.

fedorov replied:
> I know of a solution but you may not like it. In my opinion, there is no reason to have duplicate content in any columns. i.e we must create nesting.

pip package size is a precious and limited resource. Why would we want to keep bucket information in any of the columns of the main index - nested or not - if we can do static mapping with just 3 rows? Again, I do not understand the motivation.

vkt1414 replied:
Exactly my point as well that pip size is precious. But I meant in general. Collection_id, patientid, study id  and several others columns have only a fraction of unique values and there is absolutely no need to populate them in every line when we can nest anything common above series level attributes. 

I'll write a nested query and show you the savings from nesting vs the current one. But yes it will be populated in more than 3 rows for sure. 

Essentially my point is equivalent to splitting our one flatenned table to several small tables and joining just like the way you are referring to the bucket mapping.

fedorov replied:
I agree we should consider different alternatives. Let's address this outside of this PR.

fedorov replied:
I suggest we do the following:

1. make this check optional as well
2. keep this check enabled by default
3. add console log message to the user informing about what is going on (ie, "Checking validity of manifest UUIDs ...").

vkt1414 replied:
this is an internal function, we can introduce a new parameter called `validate_manifest` in `download_from_manifest` and call this internal function only when `validate_manifest`=True. Otherwise, the manifest is passed directly to _s5cmd_run.

I'll add the logging message.

vkt1414 replied:
> Otherwise, the manifest is passed directly to _s5cmd_run.

I realize it's more complicated than that. We need to parse each line regardless to get total size. I'll still pass the manifest to validator but ignore all errors when `validate_manifest= False`

fedorov commented on idc_index/index.py:
```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+        # Check if there are more than one endpoints
+        if len(merged_df["endpoint"].unique()) > 1:
+            raise ValueError(
+                "The manifest contains URLs from both AWS and GCP. Please use only one provider."
```
I may as well have a bogus bucket ID - this message will not be accurate in the general case and may be misleading.

vkt1414 replied:
checking the first url with `s5cmd ls` if endpoint is not aws. All unusual errors are captured by a generic error.

vkt1414 replied:
Reverted to printing a more specific error that catches both possibilities "Either GCS bucket path is invalid or manifest has a mix of GCS and AWS urls. If so, please use urls from one provider only"
"

fedorov commented on idc_index/index.py:
```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+            with subprocess.Popen(
+                cmd, stdout=stdout, stderr=stderr, universal_newlines=True
+            ) as process:
+                print("Total requested download size:", round(total_size, 2), " MB")
```
I suggest adding a check here that would switch to GB if the size is above 1000 MB, and TB if the size is above 1000,000 MB to improve readability of this message.

fedorov commented on idc_index/index.py:
```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+                sum(
+                    f.stat().st_size for f in Path(downloadDir).iterdir() if f.is_file()
+                )
+                - initial_size_bytes
```
I guess it should be ok if some other process downloads/copies files to the same directory at the same time - the worst is it would jump to 100% and be stuck there - right?

vkt1414 replied:
Download continues, progress bar stops and does not go over after reaching 100%. Resolving.

fedorov reviewed (changes requested):
Nice work, I think it is very close!

fedorov commented:
I submitted few small edits to your branch (I think I will want to review the docstrings separately after you move out the dev content to comments): https://github.com/vkt1414/idc-index/pull/9.

Also, can you add some logging statements to help debug this, like I added that one statement in the PR above?

fedorov commented on idc_index/index.py:
```diff
@@ -510,6 +831,7 @@ def download_from_selection(
         seriesInstanceUID=None,
+        quiet=False,
     ):
         """Download the files corresponding to the selection. The filtering will be applied in sequence (but does it matter?) by first selecting the collection(s), followed by
```
Sorry, I realized I commented on a wrong line - the comment is about this statement: "The filtering will be applied in sequence (but does it matter?) by first selecting the collection(s)".

vkt1414 replied:
I did not write this docstring.. it was already in the main branch. But yes, we could address this in another PR I believe.
https://github.com/ImagingDataCommons/idc-index/blame/main/idc_index/index.py#L514

fedorov commented:
@vkt1414 please let me know when this is ready for review!

vkt1414 reviewed (commented):
@fedorov This PR should now be ready for review again.

fedorov reviewed (changes requested):
As discussed today, there are some further optimization to be implemented in the validate function. Also, please see this PR for few comments/style changes. I have to say I did not finish the review - I only did the validate function...

vkt1414 commented:
> As discussed today, there are some further optimization to be implemented in the validate function. Also, please see this PR for few comments/style changes: [vkt1414#10](https://github.com/vkt1414/idc-index/pull/10). I have to say I did not finish the review - I only did the validate function...

Sounds good. I made the validator as efficient as I possibly could. Made tests more robust now by testing all combinations of optional parameters. Made pytest more verbose in CI to aid in troubleshooting.
